File: site/showcase/swift-notes/Sources/Notes/NotesRepository.swift
import Foundation

/// Single source of truth — Core Data locally, optional network refresh.
final class NotesRepository {
  private let store: NotesStore
  private let api: NotesAPI

  init(store: NotesStore = NotesStore(), api: NotesAPI = NotesAPI()) {
    self.store = store
    self.api = api
  }

  func fetchAll() -> [Note] {
    store.fetchAll()
  }

  func insert(_ note: Note) {
    store.insert(note)
  }

  func refreshFromNetwork() {
    let remote = api.fetchNotes()
    store.replaceAll(remote)
  }
}

File: site/showcase/android-notes/app/src/main/java/com/example/notes/NotesRepository.java
package com.example.notes;

import androidx.lifecycle.LiveData;

/**
 * Single source of truth for notes — Room locally, optional network refresh.
 */
public class NotesRepository {
  private final NoteDao noteDao;
  private final NotesApi notesApi;

  public NotesRepository(NoteDao noteDao) {
    this(noteDao, new NotesApi());
  }

  public NotesRepository(NoteDao noteDao, NotesApi notesApi) {
    this.noteDao = noteDao;
    this.notesApi = notesApi;
  }

  public LiveData<java.util.List<Note>> observeNotes() {
    return noteDao.observeAll();
  }

  public void insert(Note note) {
    noteDao.insert(note);
  }

  public void refreshFromNetwork() {
    java.util.List<Note> remote = notesApi.fetchNotes();
    noteDao.replaceAll(remote);
  }
}
